response_operations_ui/views/messages.py

```diff
@@ -154,6 +155,91 @@ def view_conversation(thread_id):
+
+    survey_list = get_grouped_surveys_list()
+
+    return render_template("secure-message/change-thread-category.html",
+                           thread=thread,
+                           thread_id=thread_id,
+                           breadcrumbs=breadcrumbs,
+                           survey_list=survey_list,
+                           form=form)
+
+
+@messages_bp.route('/threads/<thread_id>/change-category', methods=['POST'])
+@login_required
+def post_change_thread_category(thread_id):  # noqa: C901
+    thread = message_controllers.get_conversation(thread_id)
+    form = ChangeThreadCategoryForm(request.form)
+
+    if form.validate():
+        category = form.category.data
+        update_survey_id_in_messages = False
+
+        # Do this part first as a failed survey call after the thread patch could leave us in a broken state
+        if category == 'SURVEY':
+            selected_survey = form.select_survey.data
+            survey_id = survey_controllers.get_survey_id_by_short_name(selected_survey)
+            if thread['messages'][0]['survey_id'] != survey_id:
+                update_survey_id_in_messages = True
+
+        if category != thread['category']:
+            payload = {'category': category}
+            message_controllers.patch_thread(thread_id, payload)
+
+        # When the category is survey and that survey is different from what is against the messages in the thread,
+        # we need to add the survey_id to every message in the thread as the thread doesn't store that information.
+        if category == 'SURVEY' and update_survey_id_in_messages:
+            selected_survey = form.select_survey.data
+            survey_id = survey_controllers.get_survey_id_by_short_name(selected_survey)
+            if thread['messages'][0]['survey_id'] != survey_id:
+                for message in thread['messages']:
+                    message_id = message['msg_id']
+                    try:
+                        message_payload = {'survey_id': survey_id}
+                        message_controllers.patch_message(message_id, message_payload)
+                    except ApiError:
+                        # If something goes wrong with any of these calls, we'll try and revert it to what it was as
+                        # we don't want to risk any threads getting lost (i.e., A technical message being changed to
+                        # survey but the survey_id not being present anywhere)
+                        logger.error("Something went wrong updating the survey_id in a message.  Reverting thread "
+                                     "category back to what it originally was to prevent data loss",
+                                     message_id=message_id,
+                                     original_category=thread['category'])
+                        payload = {'category': thread['category']}
+                        message_controllers.patch_thread(thread_id, payload)
```

Comment:
I don't think we should be taking a rollback approach with this - if a message ID patch fails because the SM service has gone down in the meantime, the rollback will fail too and then we're in a right pickle.

I'd much rather we did this whole bit as part of the first code block at line 187 (which would also save us doubling up on getting the survey ID by short name 😃 ), eg:

```
if category == 'SURVEY':
   selected_survey = form.select_survey.data
   survey_id = survey_controllers.get_survey_id_by_short_name(selected_survey)
   if thread['messages'][0]['survey_id] != survey_id:
       # rest of this block goes here
       for message in thread['messages']:
            # blah blah blah...
            # no rollback of the thread category needed because we haven't done it yet

if category != thread['category']:
    # this block now
```

If you choose to not do this, at the very least I think we shouldn't be double calling `get_survey_id_by_short_name` and we can remove the `if thread['messages'][0]['survey_id'] != survey_id` conditional because its already been checked at the top of this block with `update_survey_id_in_messages`